### PR TITLE
[TRTLLM-11540][feat] Support rejection sampling in EAGLE3 dynamic tree

### PR DESCRIPTION
**Pull request by zhaoyangwang-nvidia**
## Summary by CodeRabbit

* **New Features**
  * Added rejection sampling for one-model speculative decoding, enabling non-greedy sampling modes (temperature, top-k, top-p) with speculative decoding.
  * Implemented dynamic tree rejection sampling for more efficient speculative token generation.
  * Added `use_rejection_sampling` configuration option to control rejection sampling behavior.

## Summary

This PR adds rejection sampling support for one-model speculative decoding, extends it to the EAGLE3 one-model dynamic-tree path, and brings the non-dynamic-tree EAGLE3 rejection verification flow under TRT-LLM instead of relying on FlashInfer. It also includes optimization/instrumentation for dynamic-tree rejection verification and adds unit/integration coverage for both dynamic-tree and non-dynamic-tree smoke paths.

## Testing

Added EAGLE3 one-model rejection sampling coverage:

- `tests/unittest/_torch/speculative/test_eagle3.py::test_llama_eagle3_rejection_sampling_modes`
- `tests/integration/defs/accuracy/test_llm_api_pytorch.py::TestLlama3_1_8BInstruct::test_eagle3_rejection_dynamic_tree_smoke[no_dynamic_tree]`
- `tests/integration/defs/accuracy/test_llm_api_pytorch.py::TestLlama3_1_8BInstruct::test_eagle3_rejection_dynamic_tree_smoke[dynamic_tree]`

The integration smoke tests are listed in:

- `tests/integration/test_lists/test-db/l0_dgx_b200.yml`
- `tests/integration/test_lists/qa/llm_function_core.txt`

### Eagle3 — Dynamic Tree ON

**Llama 3.1 8B**
- Accept rate: 0.46 → 0.53 (+15.22%)
- Accept length: 3.76 → 4.16 (+10.64%)
- Total output token rate: 641.07 → 714.16 (+11.40%)

**Qwen 3 8B**
- Accept rate: 0.58 → 0.65 (+12.07%)
- Accept length: 4.46 → 4.92 (+10.31%)
- Total output token rate: 659.43 → 709.74 (+7.63%)

**Llama 3.3 70B**
- Accept rate: 0.48 → 0.51 (+6.25%)
- Accept length: 3.87 → 4.09 (+5.68%)
- Total output token rate: 74.38 → 78.07 (+4.96%)

### Eagle3 — Dynamic Tree OFF

**Llama 3.1 8B**
- Accept rate: 0.26 → 0.29 (+11.54%)
- Accept length: 2.56 → 2.75 (+7.42%)
- Total output token rate: 585.45 → 601.45 (+2.73%)

**Qwen 3 8B**
- Accept rate: 0.31 → 0.35 (+12.90%)
- Accept length: 2.84 → 3.13 (+10.21%)
- Total output token rate: 597.38 → 617.64 (+3.39%)

**Llama 3.3 70B**
- Accept rate: 0.28 → 0.31 (+10.71%)
- Accept length: 2.71 → 2.82 (+4.06%)
- Total output token rate: 56.79 → 57.82 (+1.81%)

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Comment by zhaoyangwang-nvidia** [comment]
Hi @sunnyqgg This PR is ready, could you help to review it, thanks~

File: cpp/tensorrt_llm/kernels/speculativeDecoding/dynamicTreeKernels.cu [hunk 1]
```diff
@@ -0,0 +1,1562 @@
...
+// Two-stage top-k / top-p masking kernels
+// Mirrors the approach in invokeBatchTopKSampling (samplingTopKKernels.cu),
+// but outputs a masked logits tensor instead of sampling a token.
+// ---------------------------------------------------------------------------
+
+// Stage 1: Parallel top-k reduction across BLOCKS_PER_BEAM_ blocks per row.
+// Each block handles (vocabSize / BLOCKS_PER_BEAM_) elements and finds its
+// local top-k, writing (global_index, logit_value) pairs into the tmp buffers.
+template <typename T, int32_t BLOCK_SIZE_, int32_t BLOCKS_PER_BEAM_>
+__global__ void topKProbStage1(T const* __restrict__ logits, T* tmpLogProbs, int32_t* topKTmpIdBuf, T* topKTmpValBuf,
```
Comment (ziyixiong-nv) [inline cpp/tensorrt_llm/kernels/speculativeDecoding/dynamicTreeKernels.cu:56]: Did you do any perf analysis (e.g. ncu) for the new added kernels?

**Comment by ziyixiong-nv** [comment]
Is there any perf data on different batch sizes?

File: cpp/tensorrt_llm/kernels/speculativeDecoding/dynamicTreeKernels.cu (hunk 1 above)
Comment (zhaoyangwang-nvidia) [reply cpp/tensorrt_llm/kernels/speculativeDecoding/dynamicTreeKernels.cu:56]: I did some light perf comparisons using Nsight Systems to check the timeline and relative kernel time, but I didn’t run a detailed Nsight Compute analysis and I didn’t record exact numbers.

This kernel is essentially based on the same two-stage top-k/top-p approach as invokeBatchTopKSampling in cpp/tensorrt_llm/kernels/samplingTopKKernels.cu. I removed the actual sampling step and instead output masked logits. Among the variants ways I tried, this version was the fastest in my tests.

**Comment by zhaoyangwang-nvidia** [comment]
> Is there any perf data on different batch sizes?

The data I shared above was tested with batch 4, and so far I’ve only tested that configuration. Do you want me to add perf numbers for other batch sizes as well?

File: tests/integration/defs/accuracy/test_llm_api_pytorch.py [hunk 2]
```diff
@@ -186,6 +186,55 @@ def test_dummy_load_format(self):
             task = MMLU(self.MODEL_NAME)
             task.evaluate(llm, is_integration_test=True)
 
+    @pytest.mark.skip_less_device(4)
+    @pytest.mark.parametrize("use_dynamic_tree", [False, True],
+                             ids=["no_dynamic_tree", "dynamic_tree"])
+    def test_eagle3_rejection_dynamic_tree_smoke(self, use_dynamic_tree,
```
Comment (xinhe-nv) [inline tests/integration/defs/accuracy/test_llm_api_pytorch.py:192]: does this case support pre-blackwell?

File: tests/integration/test_lists/test-db/l0_dgx_b200.yml [hunk 3]
```diff
@@ -283,6 +283,8 @@ l0_dgx_b200:
   - accuracy/test_llm_api_pytorch.py::TestGPTOSS::test_eagle3_4gpus[v2_kv_cache-cutlass-two_model-overlap_scheduler]
   - accuracy/test_llm_api_pytorch.py::TestGPTOSS::test_eagle3_4gpus[v1_kv_cache-cutlass-two_model-no_overlap_scheduler]
   - accuracy/test_llm_api_pytorch.py::TestGPTOSS::test_eagle3_4gpus[v2_kv_cache-cutlass-two_model-no_overlap_scheduler]
+  - accuracy/test_llm_api_pytorch.py::TestLlama3_1_8BInstruct::test_eagle3_rejection_dynamic_tree_smoke[no_dynamic_tree]
```
Comment (xinhe-nv) [inline tests/integration/test_lists/test-db/l0_dgx_b200.yml]: add both of the tests into QA test list, too. Aside from that, we need to pay attention to the supported gpu

**Comment by xinhe-nv** [comment]
@coderabbitai review the code as a QA engineer.

File: tests/integration/defs/accuracy/test_llm_api_pytorch.py (hunk 2 above)
Comment (zhaoyangwang-nvidia) [reply tests/integration/defs/accuracy/test_llm_api_pytorch.py:192]: This case is not Blackwell-specific—the same test should run on pre-Blackwell GPUs as well. It is only a lightweight functional smoke (short GSM8K run with rejection sampling and optional dynamic tree), and we registered it on one CI target (DGX B200) on purpose to avoid extra CI load while still covering the path.

If QA prefers broader coverage, we can add the same two parametrized nodes to H100 (or another agreed stage) as well; otherwise we’re happy to keep it B200-only for CI scheduling. Do you want it duplicated on other platforms, or is the current single-platform setup acceptable?

**Comment by zhaoyangwang-nvidia** [comment]
For comments from coderabbitai

1. Fixed.

2. This does not apply to dynamic-tree rejection. That path does not use `SpecMetadata.draft_probs`; it uses `_draft_depth_logits_cat` and computes `draft_probs_tree` separately.

3. MTP does not need `super().prepare()` for rejection sampling because `use_rejection_sampling` is restricted to EAGLE3 by config validation.

4. CUDA and Python paths both apply the greedy one-hot override, so there is no behavior divergence there.

5. Fixed.

6. Fixed.

7. The production dynamic-tree path does not use fixed RNG defaults; it increments the persistent CUDA `seed` tensor before each rejection call.

8. Fixed.

9. The dynamic-tree `_can_use_rejection_sampling` check intentionally differs from the base path because it uses `_draft_depth_logits_cat` and verifies gen rows only.

Also added the two new integration smoke tests to `tests/integration/test_lists/qa/llm_function_core.txt`.


1. Fixed.

2. This does not apply to dynamic-tree rejection. That path does not use `SpecMetadata.draft_probs`; it uses `_draft_depth_logits_cat` and computes `draft_probs_tree` separately.

3. MTP does not need `super().prepare()` for rejection sampling because `use_rejection_sampling` is restricted to EAGLE3 by config validation.

4. CUDA and Python paths both apply the greedy one-hot override, so there is no behavior divergence there.

5. Fixed.

6. Fixed.

7. The production dynamic-tree path does not use fixed RNG defaults; it increments the persistent CUDA `seed` tensor before each rejection call.

8. Fixed.

9. The dynamic-tree `_can_use_rejection_sampling` check intentionally differs from the base path because it uses `_draft_depth_logits_cat` and verifies gen rows only.

**Review by StanleySun639 — APPROVED** [review]

**Comment by zhaoyangwang-nvidia** [comment]
Hi @sunnyqgg this PR is ready for review, please take a look.

**Comment by zhaoyangwang-nvidia** [comment]
Hi @xinhe-nv do you have more comments about this PR?

**Review by ziyixiong-nv — APPROVED** [review]

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 4]
```diff
@@ -777,6 +789,14 @@ def _get_full_general_warmup_requests(
         # Deduplicate the warmup_configs while keeping the order.
         return list(dict.fromkeys(warmup_configs))
 
+    def _get_dynamic_tree_warmup_max_batch_size(self) -> Optional[int]:
+        """Return the dynamic-tree worker batch capacity for warmup batch clamping."""
+        if (self.spec_config is None or self.is_draft_model
+                or not self.spec_config.spec_dec_mode.use_one_engine()
+                or not getattr(self.spec_config, "use_dynamic_tree", False)):
+            return None
+        return getattr(self.spec_config, "max_batch_size", None)
```
Comment (Funatiq) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py]: Please forgive my naive question, but why would we set the `max_batch_size` via `spec_config` in this case? Wouldn't it make more sense to always use the global `max_batch_size`? I don't think it's a good idea to allow configuring the `max_batch_size` in multiple ways.

Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py]: Thanks for your comments. 

You're right — having two configurable max_batch_size was confusing. The field on Eagle3DecodingConfig only existed because Eagle3OneModelDynamicTreeWorker.__init__ needs the global max_batch_size to pre-allocate batch-indexed CUDA buffers, and the worker construction site didn't have access to the global value otherwise. It was always meant to mirror the global, never tuned independently — and leaving it user-facing was risky: a smaller value would OOB on batch_idx overflow at runtime (the warmup clamps only papered over warmup, not
  real generation).

  In the latest commit I converted it to a PrivateAttr (_max_batch_size), following the same pattern as _allow_chain_drafter on DecodingBaseConfig. It's now auto-populated from the global max_batch_size in py_executor_creator as the single writer, and StrictBaseModel's extra="forbid" means users who try to set it now get an explicit ValidationError. With that invariant guaranteed structurally, I removed the three warmup/CUDA-graph clamps and the helper in model_engine.py.

Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py]: Hi @Funatiq do you have more comments about this PR?

**Review by Funatiq — APPROVED** [review]
LGTM for runtime related changes.

**Review by Superjomn — APPROVED** [review]
LGTM on the llmapi changes

**Comment by zhaoyangwang-nvidia** [comment]
Hi @NVIDIA/trt-llm-doc-owners @NVIDIA/trt-llm-torch-graph-compiler, could you please help to review this PR, thanks a lot.

File: docs/source/features/speculative-decoding.md [hunk 5]
```diff
@@ -59,7 +59,8 @@ To enable dynamic tree mode, set `use_dynamic_tree=True` on the `Eagle3DecodingC
 * `use_dynamic_tree` (`bool`): Enables dynamic tree draft generation. Mutually exclusive with `eagle_choices` (static tree).
 * `dynamic_tree_max_topK` (`int`): Maximum number of tokens to expand per node at each draft layer.
 * `max_total_draft_tokens` (`int`, optional): Total draft token budget for the tree. Must satisfy `max_draft_len <= max_total_draft_tokens <= dynamic_tree_max_topK * max_draft_len`. Defaults to `dynamic_tree_max_topK * max_draft_len` if not set.
-* `max_batch_size` (`int`): Required when `use_dynamic_tree=True` for pre-allocating dynamic tree CUDA buffers.
+
+When `use_dynamic_tree=True`, the dynamic tree CUDA buffers are pre-allocated based on the `LLM`'s `max_batch_size`; it is propagated internally and must not be passed to `Eagle3DecodingConfig` directly.
```
Comment (nv-guomingz) [inline docs/source/features/speculative-decoding.md]: which is propagated internally and must not be passed directly to Eagle3DecodingConfig.

Is it better and more clear?

**Review by nv-guomingz — APPROVED** [review]
LGTM for doc part

**Comment by zhaoyangwang-nvidia** [comment]
Hi @mikeiovine , this PR is related to rejection samples. You mentioned in a previous meeting that you’d like to take a look, so I’m tagging you here.

File: tensorrt_llm/_torch/speculative/dynamic_tree_ops.py [hunk 6]
```diff
@@ -196,3 +238,113 @@ def verify_dynamic_tree_greedy_out(
...
+                temperatures,
+                num_draft_tokens,
+                top_k,
+                top_p,
+                skip_temperature,
+                top_k_max=top_k_max,
+                skip_all_sampling_params=skip_all_sampling_params,
+            )
+
+            torch.ops.trtllm.verify_dynamic_tree_rejection_out_op(
```
Comment (hyukn) [inline tensorrt_llm/_torch/speculative/dynamic_tree_ops.py:325]: Do we have a register fake for this op? Because this is a non-compiled tree paths, it may not be necessary.

File: tensorrt_llm/_torch/speculative/eagle3_dynamic_tree.py [hunk 7]
```diff
@@ -752,21 +881,109 @@ def _sample_and_accept_dynamic_tree(
...
+        num_gens: int,
+    ) -> torch.Tensor:
+        """Build tree-position -> unique draft-prob row mapping.
+
+        For a final tree position:
+          - depth 0 children map to row 0, which stores p(.|root)
+          - deeper nodes map to row 1 + depth_bucket * K + parent_k
+        """
+        draft_prob_indices = self._draft_prob_indices_buf[:num_gens]
+        torch.ops.trtllm.build_draft_prob_indices_out_op(
```
Comment (hyukn) [inline tensorrt_llm/_torch/speculative/eagle3_dynamic_tree.py:979]: Same here.

File: tensorrt_llm/_torch/speculative/one_model_sampler.py [hunk 8]
```diff
@@ -92,6 +99,74 @@ def sampling_batch_spec_dec_one_model(
...
+    logits: torch.Tensor,
+    temperatures: torch.Tensor,
+    top_k: Optional[torch.Tensor],
+    top_p: Optional[torch.Tensor],
+    skip_temperature: bool = False,
+) -> torch.Tensor:
+    """
+    Compute probabilities from logits with temperature, top-k, and top-p applied.
+    """
+    if logits.is_cuda and hasattr(torch.ops.trtllm, "compute_probs_from_logits_op"):
```
Comment (hyukn) [inline tensorrt_llm/_torch/speculative/one_model_sampler.py]: Why use hasattr here? Is it not guaranteed to be registered?

**Review by hyukn — APPROVED** [review]
LGTM. Some minor concerns of the op registration.

File: tensorrt_llm/_torch/speculative/dynamic_tree_ops.py (hunk 6 above)
Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/speculative/dynamic_tree_ops.py:325]: Added a no-op fake impl for it.

File: tensorrt_llm/_torch/speculative/eagle3_dynamic_tree.py (hunk 7 above)
Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/speculative/eagle3_dynamic_tree.py:979]: Added a no-op fake impl for it.

**Comment by zhaoyangwang-nvidia** [comment]
Hi @sunnyqgg I pushed a small review-followup commit: added no-op fake impls for two new out-ops, removed the defensive `hasattr` guard on the CUDA custom op path, and applied a docs wording suggestion. Previous CI was fully green; could we merge without rerunning CI to reduce CI load?
ci2026-05-14 151038
